Augment `brew livecheck` with a `--resources` option to check resources

mohammadzainabbas opened the pull request:
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

This PR aims to extend the `brew livecheck` command to work for resources as well. 

---

I have added an option `-r` or `--resources` in `brew livecheck` command which do a livecheck for resources when given a particular Formulae.

_For instance_: inside `Library/Taps/homebrew/homebrew-core/Formula/influxdb.rb`, add the following livecheck block for `pkg-config-wrapper` resource:

```ruby
...
resource "pkg-config-wrapper" do
    url "https://github.com/influxdata/pkg-config/archive/refs/tags/v0.2.11.tar.gz"
    sha256 "52b22c151163dfb051fd44e7d103fc4cde6ae8ff852ffc13adeef19d21c36682"

    livecheck do
      url "https://github.com/influxdata/pkg-config/tags"
      regex(%r{href="/influxdata/pkg-config/archive/refs/tags/v(.*).tar.gz"}i)
      strategy :page_match
    end
  end
...
```
and then run the following command:

```bash
brew livecheck influxdb --resources
```

or 

```bash
brew livecheck influxdb -r
```

you will get the resources' information as well (as expected)

```txt
influxdb: 2.4.0 ==> 2.4.0
  pkg-config-wrapper: 0.2.11 ==> 0.2.12
  ui-assets: 2.4.0 ==> 2022-09-16
```

---

Also, this PR includes the implementation for resource information in the default, `--debug`, and `--json` output for resources (will be needing valuable feedback for this)

```bash
brew livecheck influxdb --resources --debug
``` 

will give us the below output

```txt
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/influxdb.rb

Formula:          influxdb
Livecheckable?:   Yes

URL (stable):     https://github.com/influxdata/influxdb.git
Strategy:         Git
Regex:            /^v?((?!9\.9\.9)\d+(?:\.\d+)+)$/i

Matched Versions:
1.8.0, 0.0.1, 0.0.2, 0.0.3, 0.0.4, 0.0.5, 0.0.6, 0.0.7, 0.0.8, 0.0.9, 0.1.0, 0.10.0, 0.10.1, 0.10.2, 0.10.3, 0.11.0, 0.11.1, 0.12.0, 0.12.1, 0.12.2, 0.13.0, 0.2.0, 0.3.0, 0.3.1, 0.3.2, 0.4.0, 0.4.1, 0.4.2, 0.4.3, 0.4.4, 0.5.0, 0.5.1, 0.5.10, 0.5.11, 0.5.12, 0.5.2, 0.5.3, 0.5.4, 0.5.5, 0.5.6, 0.5.7, 0.5.8, 0.5.9, 0.6.0, 0.6.1, 0.6.2, 0.6.3, 0.6.4, 0.6.5, 0.7.0, 0.7.1, 0.7.2, 0.7.3, 0.8.0, 0.8.1, 0.8.2, 0.8.3, 0.8.4, 0.8.5, 0.8.6, 0.8.7, 0.8.8, 0.9.0, 0.9.1, 0.9.2, 0.9.2.1, 0.9.3, 0.9.4, 0.9.4.1, 0.9.4.2, 0.9.5, 0.9.5.1, 0.9.6, 0.9.6.1, 1.0.0, 1.0.1, 1.0.2, 1.1.0, 1.1.1, 1.1.2, 1.1.3, 1.1.4, 1.1.5, 1.10.0, 1.2.0, 1.2.1, 1.2.2, 1.2.3, 1.2.4, 1.3.0, 1.3.1, 1.3.2, 1.3.3, 1.3.4, 1.3.5, 1.3.6, 1.3.7, 1.3.8, 1.3.9, 1.4.0, 1.4.1, 1.4.2, 1.4.3, 1.5.0, 1.5.1, 1.5.2, 1.5.3, 1.5.4, 1.5.5, 1.6.0, 1.6.1, 1.6.2, 1.6.3, 1.6.4, 1.6.5, 1.6.6, 1.7.0, 1.7.1, 1.7.10, 1.7.11, 1.7.2, 1.7.3, 1.7.4, 1.7.5, 1.7.6, 1.7.7, 1.7.8, 1.7.9, 1.8.1, 1.8.10, 1.8.2, 1.8.3, 1.8.4, 1.8.5, 1.8.6, 1.8.7, 1.8.8, 1.8.9, 1.9.0, 1.9.1, 1.9.2, 1.9.3, 1.9.4, 1.9.5, 1.9.6, 1.9.7, 1.9.8, 2.0.0, 2.0.1, 2.0.2, 2.0.3, 2.0.4, 2.0.5, 2.0.6, 2.0.7, 2.0.8, 2.0.9, 2.1.0, 2.1.1, 2.2.0, 2.3.0, 2.4.0


Resource:         pkg-config-wrapper
Livecheckable?:   Yes

URL:              https://github.com/influxdata/pkg-config/tags
Strategy:         PageMatch
Regex:            /href="\/influxdata\/pkg-config\/archive\/refs\/tags\/v(.*).tar.gz"/i

Matched Versions:
0.2.12, 0.2.11, 0.2.10, 0.2.9, 0.2.8, 0.2.7, 0.2.6, 0.2.5, 0.2.4, 0.2.3


Resource:         ui-assets
Livecheckable?:   No

URL:              https://github.com/influxdata/ui/releases/download/OSS-Master/build.tar.gz
URL (processed):  https://github.com/influxdata/ui.git
Strategy:         Git

Matched Versions:
r, 2.1.0, 2.1.1, 2.1.2, 2022-09-01, 2022-09-16, 9.9.9.test, 2.0.5, 2.0.7, 2.0.8, 2.0.9

influxdb: 2.4.0 ==> 2.4.0
  pkg-config-wrapper: 0.2.11 ==> 0.2.12
  ui-assets: 2.4.0 ==> 2022-09-16
```

and 

```bash
brew livecheck influxdb --resources --json
```

will output the following

```json
[
  {
    "formula": "influxdb",
    "version": {
      "current": "2.4.0",
      "latest": "2.4.0",
      "outdated": false,
      "newer_than_upstream": false
    },
    "resources": [
      {
        "resource": "pkg-config-wrapper",
        "version": {
          "current": "0.2.11",
          "latest": "0.2.12",
          "outdated": true,
          "newer_than_upstream": false
        }
      },
      {
        "resource": "ui-assets",
        "version": {
          "current": "2.4.0",
          "latest": "2022-09-16",
          "outdated": true,
          "newer_than_upstream": false
        }
      }
    ]
  }
]
```

---

Kindly, do let me know if there's something missing or any further improvement that we can do here.

nandahkrishna reviewed (commented):
I've left a brief review containing the most important points I discussed during yesterday's meeting with Mohammad. There's definitely scope for further improvement, especially with respect to the output format/presentation. I've also likely missed other important changes in this review, but I thought I'd let Mohammad work through the larger issues we identified first before providing more fine-grained feedback.

samford commented on Library/Homebrew/livecheck/livecheck.rb:
```diff
@@ -561,6 +678,214 @@ def use_homebrew_curl?(formula_or_cask, url)
+        puts <<~EOS
+
+        EOS
+        odebug "Resource:          #{resource_name(resource, full_name: full_name)}"
+        odebug "Livecheckable?:    #{has_livecheckable ? "Yes" : "No"}"
```
```suggestion
        puts "Resource:          #{resource_name(resource, full_name: full_name)}"
        puts "Livecheckable?:    #{has_livecheckable ? "Yes" : "No"}"
```

This output should align with the existing style seen in `#latest_version` (namely, using `puts` here instead of `odebug`).

mohammadzainabbas replied:
Okay ✅

samford commented on Library/Homebrew/livecheck/livecheck.rb:
```diff
@@ -394,36 +472,69 @@ def formula_name(formula, full_name: false)
+          info[:version][:latest]
+        end
+
+        puts "-- #{resource_s}: #{current_s} ==> #{latest_s}"
```
```suggestion
        puts "  #{resource_s}: #{current_s} ==> #{latest_s}"
```

I tend to agree with Nanda's suggestion on Slack of simply indenting resources by two spaces, unless someone has a more compelling idea.

mohammadzainabbas replied:
Done ✅

samford reviewed (commented):
I'll take some time to go through this over the weekend (I've been busy with work this week) but here are a couple small tweaks to the output in the interim time. I haven't tested this yet but the example JSON in the PR description is in line with what I was imagining.

mohammadzainabbas commented on Library/Homebrew/livecheck/livecheck.rb:
```diff
@@ -282,6 +283,36 @@ def run_checks(
+              { name: resource.name, version: resource.version, livecheckable: resource.livecheckable? }
+            end
+
+            resource_version_info = resource_version(
```
Done

mohammadzainabbas commented on Library/Homebrew/livecheck/livecheck.rb:
```diff
@@ -334,7 +367,50 @@ def run_checks(
+            is_newer_than_upstream = current > latest
+            is_outdated = (current != latest) && !is_newer_than_upstream
+
+            info = {}
```
Okay, done.

mohammadzainabbas commented on Library/Homebrew/livecheck/livecheck.rb:
```diff
@@ -394,36 +472,69 @@ def formula_name(formula, full_name: false)
 
+    # Formats and prints the livecheck result for a resource (for a given Formula or Cask).
+    sig { params(resources_info: Array(Hash), verbose: T::Boolean, ambiguous_cask: T::Boolean).void }
+    def print_latest_resource_version(resources_info, verbose:, ambiguous_cask: false)
```
Refactored `print_latest_version` to work for a single resource as well.

mohammadzainabbas commented on Library/Homebrew/livecheck/livecheck.rb:
```diff
@@ -282,6 +283,36 @@ def run_checks(
+          has_resources = formula_or_cask.resources.any?
+
+          # Only check current and latest versions if we have resources to check against
+          if has_resources
```
Actually, I was using `has_resources` to check if we have any resources to print output for resources or not. But anyway, I have refracted the code as suggested.

samford commented on Library/Homebrew/livecheck/livecheck.rb:
```diff
@@ -394,42 +461,54 @@ def formula_name(formula, full_name: false)
+      package_or_resource_s = "  " if check_resource
+      package_or_resource_s += "#{Tty.blue}#{info[:formula] || info[:cask] || info[:resource]}#{Tty.reset}"
+      package_or_resource_s += " (cask)" if ambiguous_cask && !check_resource
+      package_or_resource_s += " (guessed)" if !info[:meta][:livecheckable] && !check_resource && verbose
```
`info[:meta][:livecheckable]` is currently giving me an `undefined method `[]' for nil:NilClass` error when testing `influxdb` using the instructions in the PR description. This occurs when livecheck attempts to print information for a resource:

```
influxdb: 2.3.0 ==> 2.3.0
Error: influxdb: undefined method `[]' for nil:NilClass
```

To resolve this, you'll need to look into why `info[:meta]` is `nil` for a resource at this point. This may also explain why the `livecheckable` field is missing for resource objects in the verbose JSON output.

mohammadzainabbas replied:
Fixed the issue ✅

samford commented on Library/Homebrew/livecheck/livecheck.rb:
```diff
@@ -561,6 +644,216 @@ def use_homebrew_curl?(formula_or_cask, url)
+      has_livecheckable = resource.livecheckable?
+
+      if debug
+        puts <<~EOS
+
+        EOS
```
```suggestion
        puts "\n\n"
```

If you're simply looking to print a blank line, calling `puts` without an argument will accomplish this (you can see this approach used in `#latest_version`).

That said, what do we think about using a double blank line before resources to better separate the debug output? With a single blank line, resource output coming after the formula output (or another resource) tends to blend together and I feel like the extra blank line helps to keep it intelligible.

mohammadzainabbas replied:
I tend to agree with you on this one.

samford commented on Library/Homebrew/livecheck/livecheck.rb:
```diff
@@ -334,7 +360,46 @@ def run_checks(
         end
 
-        print_latest_version(info, verbose: verbose, ambiguous_cask: ambiguous_casks.include?(formula_or_cask))
+        print_latest_version(info, verbose: verbose, ambiguous_cask: ambiguous_casks.include?(formula_or_cask), check_resource: false)
```
Up to this point, the output from `#print_latest_version` has come directly after the debug output without any separation:

```
Matched Versions:
1.8.0, 0.0.1, 0.0.2, 0.0.3, 0.0.4, 0.0.5, 0.0.6, 0.0.7, 0.0.8, 0.0.9, 0.1.0, ..., 2.0.0, 2.0.1, 2.0.2, 2.0.3, 2.0.4, 2.0.5, 2.0.6, 2.0.7, 2.0.8, 2.0.9, 2.1.0, 2.1.1, 2.2.0, 2.3.0
influxdb: 2.3.0 ==> 2.3.0
```

However, now that we may be printing multiple lines when checking resource versions, this can start to look a bit cramped:

```
...

Resource:          pkg-config-wrapper
Livecheckable?:    Yes

URL:              https://github.com/influxdata/pkg-config/tags
Strategy:         PageMatch
Regex:            /href="\/influxdata\/pkg-config\/archive\/refs\/tags\/v(.*).tar.gz"/i

Matched Versions:
0.2.12, 0.2.11, 0.2.10, 0.2.9, 0.2.8, 0.2.7, 0.2.6, 0.2.5, 0.2.4, 0.2.3

Resource:          ui-assets
Livecheckable?:    No
influxdb: 2.3.0 ==> 2.3.0
  pkg-config-wrapper: 0.2.11 ==> 0.2.12
  ui-assets: 2.1.2 ==> 2.1.2
```

There may be something to be said for printing a blank line before the output from `#print_latest_version` in the debug output:

```suggestion
        puts if debug
        print_latest_version(info, verbose: verbose, ambiguous_cask: ambiguous_casks.include?(formula_or_cask), check_resource: false)
```

This ends up looking like:

```
...

Resource:          ui-assets
Livecheckable?:    No

influxdb: 2.3.0 ==> 2.3.0
  pkg-config-wrapper: 0.2.11 ==> 0.2.12
  ui-assets: 2.1.2 ==> 2.1.2
```

mohammadzainabbas replied:
Totally agreed! It was missing for Formulae/Cask as well. Initially, I added a blank space there, but later removed it in order to not change the debug output for Formulae/Cask.

samford reviewed (commented):
I'm still working on a proper review of this but I'll start with errors that I encountered in manual testing and areas where we may further improve the output.

In general, please make sure that you're running `brew style` (and `brew typecheck`) regularly and resolving any warnings/errors. At the moment, there are a number of `brew style` errors that need to be addressed. We may end up having to disable the `BlockLength` and `CyclomaticComplexity` RuboCops in places (until I dig into refactoring after GSoC is over) but we can see how those fare when this PR is closer to merging.

mohammadzainabbas commented on Library/Homebrew/livecheck/livecheck.rb:
```diff
@@ -394,42 +461,54 @@ def formula_name(formula, full_name: false)
+    # Formats and prints the livecheck result for a formula/cask/resource.
+    sig { params(info: Hash, verbose: T::Boolean, ambiguous_cask: T::Boolean, check_resource: T::Boolean).void }
+    def print_latest_version(info, verbose:, ambiguous_cask: false, check_resource: false)
+      package_or_resource_s = "  " if check_resource
```
Oops, my bad. I missed it. Fixed the issue ✅

nandahkrishna commented on Library/Homebrew/livecheck/livecheck.rb:
```diff
@@ -561,6 +646,204 @@ def use_homebrew_curl?(formula_or_cask, url)
+        },
+      }
+
+      # Check resources with livecheck block (will be updated in the future)
+      if has_livecheckable
```
As discussed earlier on Slack, we should check _all_ resources, not just the ones with `livecheck` blocks.

mohammadzainabbas replied:
Done ✅

nandahkrishna commented on Library/Homebrew/livecheck/livecheck.rb:
```diff
@@ -561,6 +646,204 @@ def use_homebrew_curl?(formula_or_cask, url)
+      if debug
+        puts "\n\n"
+        puts "Resource:          #{resource_name(resource, full_name: full_name)}"
+        puts "Livecheckable?:    #{has_livecheckable ? "Yes" : "No"}"
```
```suggestion
        puts "Livecheckable?:    #{resource.livecheckable? ? "Yes" : "No"}"
```

nandahkrishna commented on Library/Homebrew/livecheck/livecheck.rb:
```diff
@@ -561,6 +646,204 @@ def use_homebrew_curl?(formula_or_cask, url)
+      verbose: false,
+      debug: false
+    )
+      has_livecheckable = resource.livecheckable?
+
```
```suggestion
```

nandahkrishna reviewed (commented):
I've added a brief review here with some things that caught my eye, I might have more suggestions to add after some testing.

mohammadzainabbas commented:
I have refactored the code for _resources_ in `livecheck` command. Now, it works for resources without `livecheck` block (as it should). I have also updated code to get `meta` information as well (when `--json --verbose` flags are given for resources). i.e:

```bash
brew livecheck influxdb --resources --json --verbose
```
will now print the following output:

```json
[
  {
    "formula": "influxdb",
    "version": {
      "current": "2.3.0",
      "latest": "2.3.0",
      "outdated": false,
      "newer_than_upstream": false
    },
    "meta": {
      "livecheckable": true,
      "url": {
        "symbol": "stable",
        "original": "https://github.com/influxdata/influxdb.git"
      },
      "strategy": "Git",
      "strategies": [
        "Git",
        "PageMatch"
      ],
      "regex": "/^v?((?!9\\.9\\.9)\\d+(?:\\.\\d+)+)$/i"
    },
    "resources": [
      {
        "resource": "pkg-config-wrapper",
        "version": {
          "current": "0.2.11",
          "latest": "0.2.12",
          "newer_than_upstream": false,
          "outdated": true
        },
        "meta": {
          "url": {
            "original": "https://github.com/influxdata/pkg-config/tags"
          },
          "livecheckable": "Yes",
          "livecheck": {
            "url": {
              "original": "https://github.com/influxdata/pkg-config/tags"
            },
            "strategy": "PageMatch",
            "strategies": [
              "PageMatch"
            ],
            "regex": "/href=\"\\/influxdata\\/pkg-config\\/archive\\/refs\\/tags\\/v(.*).tar.gz\"/i"
          }
        }
      },
      {
        "resource": "ui-assets",
        "version": {
          "current": "2.1.2",
          "latest": "9.9.9.test",
          "newer_than_upstream": false,
          "outdated": true
        },
        "meta": {
          "url": {
            "original": "https://github.com/influxdata/ui/releases/download/OSS-2.1.2/build.tar.gz",
            "processed": "https://github.com/influxdata/ui.git"
          },
          "livecheckable": "No"
        }
      }
    ]
  }
]
```

Also, I have updated `bash`, `zsh` and `fish` completions for `brew livecheck` command to incorporate `--resources` flag. 

Now, I am looking into the tests which needs to be updated due to this update in `brew livecheck` command.

mohammadzainabbas commented:
> Now, I am looking into the tests which needs to be updated due to this update in `brew livecheck` command.

**Update**: I have updated the tests for `brew livecheck` (for the methods that I modified during this PR), and verified locally by running:

```bash
brew tests --only=livecheck/livecheck
```
and 

```bash
brew tests --only=livecheck/livecheck_version
```

nandahkrishna reviewed (commented):
We should probably keep the documentation changes for the next PR, it's getting too big to review efficiently with all these changes. In my opinion, we should just focus on getting `livecheck` to work for resources and add necessary tests in this PR.

mohammadzainabbas commented:
As per my discussion with @nandahkrishna, I have removed the usage of `use_homebrew_curl` for resources (in `resource_version` method) because in `Resource` class, `:url` is of type `String` and we can't do something like `resource.send(:url).using == :homebrew_curl` (unlike for e.g: when defined, in `Formula`, `:stable` or `:head` are of type `SoftwareSpec`, which enable us to do something like `formula.send(:stable).using == :homebrew_curl`). 

I think for the time being, livecheck for resources can work without `use_homebrew_curl`. If needed, we may need to extend `Resource` class in the future to accommodate this functionality.

Do let me know if you have a better suggestion for this.

nandahkrishna commented:
I'm taking some time to test this and go through the code in detail, so I'll hopefully post my comments by tonight (ET).

nandahkrishna commented on Library/Homebrew/livecheck/livecheck.rb:
```diff
@@ -268,7 +269,6 @@ def run_checks(
         end
 
         current_str = current.to_s
-        current = LivecheckVersion.create(formula_or_cask, current)
```
I don't think this should be removed – is there any particular reason to do so?

mohammadzainabbas replied:
Oops! my bad. I removed it because `brew typecheck` was giving me a warning here, but I guess it was before all the previous changes that I have done. Now, there's no issue with it.

nandahkrishna commented on Library/Homebrew/livecheck/livecheck.rb:
```diff
@@ -297,7 +313,6 @@ def run_checks(
         end
 
         latest_str = latest.to_s
-        latest = LivecheckVersion.create(formula_or_cask, latest)
```
Similarly, this shouldn't be removed either.

nandahkrishna commented on Library/Homebrew/livecheck/livecheck.rb:
```diff
@@ -282,6 +282,22 @@ def run_checks(
+        check_for_resources = check_resources && formula_or_cask.is_a?(Formula) && formula_or_cask.resources.present?
+        if check_for_resources
+          resource_version_info = formula_or_cask.resources.map do |resource|
+            resource_info = resource_version(
+              resource,
+              json:      json,
+              full_name: use_full_name,
+              verbose:   verbose,
+              debug:     debug,
+            )
+            resource_info
```
```suggestion
            resource_version(
              resource,
              json:      json,
              full_name: use_full_name,
              verbose:   verbose,
              debug:     debug,
            )
```
It might be sufficient to do just this since the `resource_info` variable is practically useless – it's just used to return the value that was assigned to it in the previous statement.

nandahkrishna commented on Library/Homebrew/livecheck/livecheck.rb:
```diff
@@ -532,36 +571,222 @@ def preprocess_url(url)
+
+      if debug
+        puts "\n\n"
+        puts "Resource:          #{resource.name}"
+        puts "Livecheckable?:    #{has_livecheckable ? "Yes" : "No"}"
```
```suggestion
        puts "Resource:         #{resource.name}"
        puts "Livecheckable?:   #{has_livecheckable ? "Yes" : "No"}"
```
For consistency with the formula debug output; the resource debug output wasn't properly aligned.

nandahkrishna commented on Library/Homebrew/livecheck/livecheck.rb:
```diff
@@ -532,36 +571,222 @@ def preprocess_url(url)
+        debug:     T::Boolean,
+      ).returns(T.nilable(Hash))
+    }
+    def resource_version(
```
You do not seem to have accounted for `livecheck` blocks in resources that contain `skip` – you should ensure that this behaviour is supported and identical to what happens with Formulae and Casks.

samford replied:
I also noticed that a resource `livecheck` block using `#skip` didn't work when I tested it, so I second this. At least at the moment, I believe we only need to concern ourselves with `skip` `livecheck` blocks for resources but, to achieve this, we will need to make changes to the `Livecheck::SkipConditions` module.

Off the top of my head:

* The `#formula_or_cask_skip` method should be updated to handle resources and it would probably make sense to rename it to `#package_or_resource_skip` (to align with the other changes here).
* We will need a `RESOURCE_CHECKS` array (like the existing ones for formulae and casks) that simply includes `:package_or_resource_skip`.
* `#skip_information`/`#referenced_skip_information`/`#print_skip_information` will need to be updated to handle resources as well.

If you have any questions about this while looking through the `SkipConditions` module, feel free to ask (e.g., on Slack).

nandahkrishna commented on Library/Homebrew/livecheck/livecheck.rb:
```diff
@@ -282,6 +282,22 @@ def run_checks(
+          end
+        end
+
         if latest.blank?
```
It seems like the code to handle a blank latest version for resources is not present. This situation arises when a latest version could not be found and must be implemented just like for formulae. You might have to modify `resource_version` and maybe even `status_hash` to accommodate this behaviour.

You should also keep in mind that a formula's livecheck may error out with the "unable to get versions" message, but the resource results should still be presented if possible. This would mean you'll have to rearrange/modify the code that handles a formula's blank latest version and ensure that the resource information is also output, in addition to the `status_hash` result.

nandahkrishna reviewed (commented):
Here are some issues I found with the implementation so far, and related suggestions. I'm also pinging Sam to review this when they can, as they may be able to spot something that I've missed in my review.

samford commented:
> I'm also pinging Sam to review this when they can, as they may be able to spot something that I've missed in my review.

I ran out of time tonight but I'll try to take another pass through this tomorrow evening (EDT).

samford commented on Library/Homebrew/livecheck/livecheck.rb:
```diff
@@ -396,40 +427,44 @@ def formula_name(formula, full_name: false)
+      package_or_resource_s = resource ? "  " : ""
+      package_or_resource_s += "#{Tty.blue}#{info[:formula] || info[:cask] || info[:resource]}#{Tty.reset}"
+      package_or_resource_s += " (cask)" if ambiguous_cask
+      package_or_resource_s += " (guessed)" if verbose && !resource && !info[:meta][:livecheckable]
```
```suggestion
      package_or_resource_s += " (guessed)" if verbose && !info[:meta][:livecheckable]
```

The behavior for resources should be the same as formulae in terms of the `(guessed)` output. That said, this suggested change will error because the `#resource_version` method isn't providing all the data we need in its return value. Basically, `#resource_version` should always return all of the data (since `#print_latest_version` needs `:meta` information) and `#run_checks` should take care of removing `:meta` from the resource hashes when the `--json` flag is set but the `--verbose` flag is not.

You'll need to remove a few early returns in `#resource_version` that are artificially restricting the returned data and add additional logic in `#run_checks` that mimics the existing `info.except!(:meta) unless verbose` logic (but applying to each resource hash in the resources array, of course).

If this isn't clear, take a look at the return value from `#latest_version` and how it's handled in `#run_checks`.

mohammadzainabbas replied:
Make sense. Done ✅

samford commented on Library/Homebrew/livecheck/livecheck.rb:
```diff
@@ -532,36 +572,222 @@ def preprocess_url(url)
+
+        resource_version_info[:meta] = { url: { original: original_url } }
+        resource_version_info[:meta][:url][:processed] = url if url != original_url
+        resource_version_info[:meta][:livecheckable] = has_livecheckable ? "Yes" : "No"
```
```suggestion
        resource_version_info[:meta][:livecheckable] = has_livecheckable
```

This value should align with the format of the existing JSON output. That is to say, this should be a boolean, not a "Yes"/"No" string. [A string value here may also cause some logic in `#print_latest_version` to work in an unexpected manner, as a string would be "truthy" value.]

mohammadzainabbas replied:
Oh yeah! you are right. I've made the suggested changes.

samford reviewed (commented):
I'm still working on going through this but it's late over here and I need to crawl in bed, so I'm going to try to pick this up Monday evening. For the moment, here are a few comments that should give you something to work on in the interim time. Past that, I second Nanda's comment about `skip` and I'll add a follow-up comment there.

mohammadzainabbas commented:
After verifying and testing locally the changes suggested by @nandahkrishna and @samford, I have updated this PR. Now, we account for `skip` `livecheck` blocks for resources.

For instance: inside Library/Taps/homebrew/homebrew-core/Formula/influxdb.rb, add the following _livecheck_ block for `pkg-config-wrapper` resource:

```ruby
...
  resource "pkg-config-wrapper" do
    url "https://github.com/influxdata/pkg-config/archive/refs/tags/v0.2.11.tar.gz"
    sha256 "52b22c151163dfb051fd44e7d103fc4cde6ae8ff852ffc13adeef19d21c36682"

    livecheck do
      skip "No longer actively developed"
    end
  end
...
```

and then run the following command:

```bash
brew livecheck influxdb -r
```

you will get the see the `SkipConditions.print_skip_information` will work as expected:

```txt
influxdb: 2.4.0 ==> 2.4.0
  pkg-config-wrapper: skipped - No longer actively developed
  ui-assets: 2.4.0 ==> 2022-09-01
```

and after running:

```bash
brew livecheck influxdb -rj
```

you will get the `json` output in the following manner:

```json
[
  {
    "formula": "influxdb",
    "version": {
      "current": "2.4.0",
      "latest": "2.4.0",
      "outdated": false,
      "newer_than_upstream": false
    },
    "resources": [
      {
        "resource": "pkg-config-wrapper",
        "status": "skipped",
        "messages": [
          "No longer actively developed"
        ]
      },
      {
        "resource": "ui-assets",
        "version": {
          "current": "2.4.0",
          "latest": "2022-09-01",
          "newer_than_upstream": false,
          "outdated": true
        }
      }
    ]
  }
]
```
---

Also as suggested, if the formula somehow results in error when the `-r` resource flag is given, it will still show the output for resources:

```txt
Error: influxdb: Unable to get versions
  pkg-config-wrapper: 0.2.11 ==> 0.2.12
  ui-assets: 2.4.0 ==> 2022-09-01
```

and the `json` output is:

```json
[
  {
    "formula": "influxdb",
    "status": "error",
    "messages": [
      "Unable to get versions"
    ],
    "meta": {
      "livecheckable": true
    },
    "resources": [
      {
        "resource": "pkg-config-wrapper",
        "version": {
          "current": "0.2.11",
          "latest": "0.2.12",
          "newer_than_upstream": false,
          "outdated": true
        },
        "meta": {
          "url": {
            "original": "https://github.com/influxdata/pkg-config/tags"
          },
          "livecheckable": true,
          "livecheck": {
            "url": {
              "original": "https://github.com/influxdata/pkg-config/tags"
            },
            "strategy": "PageMatch",
            "strategies": [
              "PageMatch"
            ],
            "regex": "/href=\"\\/influxdata\\/pkg-config\\/archive\\/refs\\/tags\\/v(.*).tar.gz\"/i"
          }
        }
      },
      {
        "resource": "ui-assets",
        "version": {
          "current": "2.4.0",
          "latest": "2022-09-01",
          "newer_than_upstream": false,
          "outdated": true
        },
        "meta": {
          "url": {
            "original": "https://github.com/influxdata/ui/releases/download/OSS-Master/build.tar.gz",
            "processed": "https://github.com/influxdata/ui.git"
          },
          "livecheckable": false
        }
      }
    ]
  }
]
```

---

Lastly, if the resource latest version is not found (if the latest version is blank for resource):

```txt
influxdb: 2.4.0 ==> 2.4.0
  pkg-config-wrapper: 0.2.11 ==> 0.2.12
  ui-assets: Unable to get versions
```

and the `json` output is:

```json
[
  {
    "formula": "influxdb",
    "version": {
      "current": "2.4.0",
      "latest": "2.4.0",
      "outdated": false,
      "newer_than_upstream": false
    },
    "resources": [
      {
        "resource": "pkg-config-wrapper",
        "version": {
          "current": "0.2.11",
          "latest": "0.2.12",
          "newer_than_upstream": false,
          "outdated": true
        }
      },
      {
        "resource": "ui-assets",
        "status": "error",
        "messages": [
          "Unable to get versions"
        ]
      }
    ]
  }
]
```

---

There is only one issue that I am currently running into, and it's about _Cyclomatic complexity_ for `run_checks` method. I am not sure how I can reduce the complexity any further since I have tried to minimise the checks and conditions wherever possible.

```txt
Offenses:

Library/Homebrew/livecheck/livecheck.rb:173:5: C: Cyclomatic complexity for run_checks is too high. [83/80]
    def run_checks( ...
    ^^^^^^^^^^^^^^^

1110 files inspected, 1 offense detected
```


Kindly, do let me know if there is something missing or any further improvement that we can do here.

nandahkrishna commented on Library/Homebrew/livecheck/livecheck.rb:
```diff
@@ -289,7 +307,11 @@ def run_checks(
 
           next version_info if version_info.is_a?(Hash) && version_info[:status] && version_info[:messages]
```
I wonder if we should modify `version_info` to include resource information as well (though I haven't tested a scenario where this condition is true).

mohammadzainabbas replied:
I was thinking of including resource information to `version_info` too but I was not sure about the scenario in which this line execute, since the only time 

```ruby
if version_info.is_a?(Hash) && version_info[:status] && version_info[:messages]
```
evaluates to `true` is when `latest_version` method returns a `Hash` from `status_hash` method (check below):

```ruby
...
        if messages.is_a?(Array) && match_version_map.blank?
          puts messages unless json
          next if i + 1 < urls.length

          return status_hash(formula_or_cask, "error", messages, full_name: full_name, verbose: verbose)
        end
...
```

However, I could include resource information to `version_info` if that's what you suggest

nandahkrishna commented on Library/Homebrew/livecheck/livecheck.rb:
```diff
@@ -289,7 +307,11 @@ def run_checks(
+                                                                                 verbose:   verbose)
+          latest_info[:resources] = resource_version_info if check_for_resources
+
+          next latest_info
```
The `meta` key is present even when you don't pass `--verbose`.
```
➜ brew lc -r --json influxdb
[
  {
    "formula": "influxdb",
    "status": "error",
    "messages": [
      "Unable to get versions"
    ],
    "meta": {
      "livecheckable": true
    },
    "resources": [
      {
        "resource": "pkg-config-wrapper",
        "version": {
          "current": "0.2.11",
          "latest": "0.2.12",
          "newer_than_upstream": false,
          "outdated": true
        },
        "meta": {
          "url": {
            "original": "https://github.com/influxdata/pkg-config/archive/refs/tags/v0.2.11.tar.gz",
            "processed": "https://github.com/influxdata/pkg-config.git"
          },
          "livecheckable": false
        }
      },
      {
        "resource": "ui-assets",
        "version": {
          "current": "2.4.0",
          "latest": "2022-09-16",
          "newer_than_upstream": false,
          "outdated": true
        },
        "meta": {
          "url": {
            "original": "https://github.com/influxdata/ui/releases/download/OSS-Master/build.tar.gz",
            "processed": "https://github.com/influxdata/ui.git"
          },
          "livecheckable": false
        }
      }
    ]
  }
]
```

mohammadzainabbas replied:
Are you sure that you are getting this output ? I ran the same command:

```bash
brew lc -r --json influxdb
```

and got the expected output

```json
[
  {
    "formula": "influxdb",
    "version": {
      "current": "2.4.0",
      "latest": "2.4.0",
      "outdated": false,
      "newer_than_upstream": false
    },
    "resources": [
      {
        "resource": "pkg-config-wrapper",
        "version": {
          "current": "0.2.11",
          "latest": "0.2.12",
          "newer_than_upstream": false,
          "outdated": true
        }
      },
      {
        "resource": "ui-assets",
        "version": {
          "current": "2.4.0",
          "latest": "2022-09-16",
          "newer_than_upstream": false,
          "outdated": true
        }
      }
    ]
  }
]

```

mohammadzainabbas replied:
> This occurs when you modify the `livecheck` block of the formula to raise an error. I modified `influxdb`'s `livecheck` to give an "Unable to get versions" message, as you can see in the output.

Fixed ✅ Thanks for pointing it out.

nandahkrishna reviewed (commented):
I tested this PR on a couple of formulae like `go` and I'm not sure we're quite there yet. Here's the output:
```
➜ brew lc -r go
go: 1.19.1 ==> 1.19.1
Error: go: undefined method `[]' for nil:NilClass
Error: undefined method `[]' for nil:NilClass
/usr/local/Homebrew/Library/Homebrew/livecheck/livecheck.rb:467:in `print_latest_version'
/usr/local/Homebrew/Library/Homebrew/livecheck/livecheck.rb:489:in `block in print_resources_info'
/usr/local/Homebrew/Library/Homebrew/livecheck/livecheck.rb:485:in `each'
/usr/local/Homebrew/Library/Homebrew/livecheck/livecheck.rb:485:in `print_resources_info'
/usr/local/Homebrew/Library/Homebrew/livecheck/livecheck.rb:380:in `rescue in block in run_checks'
/usr/local/Homebrew/Library/Homebrew/livecheck/livecheck.rb:219:in `block in run_checks'
/usr/local/Homebrew/Library/Homebrew/livecheck/livecheck.rb:218:in `map'
/usr/local/Homebrew/Library/Homebrew/livecheck/livecheck.rb:218:in `with_index'
/usr/local/Homebrew/Library/Homebrew/livecheck/livecheck.rb:218:in `run_checks'
/usr/local/Homebrew/Library/Homebrew/dev-cmd/livecheck.rb:124:in `livecheck'
/usr/local/Homebrew/Library/Homebrew/brew.rb:95:in `<main>'
➜ brew lc -r --json go
[
  {
    "formula": "go",
    "version": {
      "current": "1.19.1",
      "latest": "1.19.1",
      "outdated": false,
      "newer_than_upstream": false
    },
    "resources": [
      {
      }
    ]
  }
]
```
These errors should be handled properly so that we get an `Unable to get versions` message in such situations (and the JSON output should contain proper status/error messages for the resources). I was able to reproduce this with `influxdb` by adding a `livecheck` block with an invalid/unreachable URL, for example.

I also have some additional comments.

mohammadzainabbas commented:
> I tested this PR on a couple of formulae like `go` and I'm not sure we're quite there yet. Here's the output:

Upon debugging, this happened when we don't have any `strategy` for a Formula's resource. For e.g: For `go`'s resource `gobootstrap` (I ran the following):

```bash
brew lc go -rd
```
you will see that 

```txt
...
Resource:         gobootstrap
Livecheckable?:   No

URL:              https://storage.googleapis.com/golang/go1.16.darwin-amd64.tar.gz
Strategy:         None
...
```
this tells that we don't have any strategy for `gobootstrap` resource.

However, I have made the changes and now you should get an expected output:

```bash
brew lc go -r
```

will result in 

```txt
go: 1.19.1 ==> 1.19.1
  gobootstrap: Unable to get versions
```

and 

```bash
brew lc go -rj
```

will yield

```json
[
  {
    "formula": "go",
    "version": {
      "current": "1.19.1",
      "latest": "1.19.1",
      "outdated": false,
      "newer_than_upstream": false
    },
    "resources": [
      {
        "resource": "gobootstrap",
        "status": "error",
        "messages": [
          "Unable to get versions"
        ]
      }
    ]
  }
]
```

---

> I was able to reproduce this with influxdb by adding a livecheck block with an invalid/unreachable URL, or a regex that doesn't match any version text (e.g. /(alpha-beta-gamma-delta)/i).

This was another issue, but now it's being handled as well.

nandahkrishna reviewed (commented):
The code mostly seems fine to me though it definitely needs to be refined (in subsequent PRs). The feature works as intended, but there is one small problem: an error in the resource check (e.g. `Unable to get versions`) does not affect the status code of the `brew livecheck` command. This should be handled properly but we can take care of it in follow-up PRs too.

I will wait until Sam has provided their feedback for a (hopefully) final round of changes before we merge this.

samford reviewed (commented):
I managed to take a pass through everything and I tried to limit the feedback to only what was essential (I've stashed other changes for a follow-up PR) but it still ended up being a bit more than expected. I feel like most of this is straightforward but there are a couple areas that will take a little work.

I would like to review the changes that are made in response to this feedback before this is merged but I believe these should be the last notable changes needed here.

mohammadzainabbas commented on Library/Homebrew/livecheck/livecheck.rb:
```diff
@@ -344,13 +384,14 @@ def run_checks(
           name += " (cask)" if ambiguous_casks.include?(formula_or_cask)
 
           onoe "#{Tty.blue}#{name}#{Tty.reset}: #{e}"
           $stderr.puts e.backtrace if debug && !e.is_a?(Livecheck::Error)
           nil
         end
+        print_resources_info(resource_version_info, verbose: verbose) if check_for_resources
```
Oh got it. Done ✅

samford replied:
> we can use `:status` to identify if the resource was skipped or was it due to some failed state, and mark `Homebrew.failed = true` in that case.

Yes, checking for a `:status` value of `"error"` was what I had in mind. However, this should arguably be handled somewhere after `#resource_version` is called in `#run_checks`, not in `#print_resources_info` (where setting `Homebrew.failed` would be an unexpected side effect).

The if/else after the `#resource_version` call also creates an error status hash, so you'll have to incorporate `Homebrew.failed = true` in a way that also handles that error status hash while still returning the value from the if/else (it's implicitly returned at the moment). If this explanation isn't entirely clear or you need additional clarification, just let me know.

mohammadzainabbas replied:
> However, this should arguably be handled somewhere after `#resource_version` is called in `#run_checks`

Is this what you had in mind ?

```ruby
check_for_resources = check_resources && formula_or_cask.is_a?(Formula) && formula_or_cask.resources.present?
if check_for_resources

  resource_version_info = formula_or_cask.resources.map do |resource|
    res_skip_info ||= SkipConditions.skip_information(resource, verbose: verbose)
    if res_skip_info.present?
      res_skip_info
    else
      res_version_info = resource_version(
        resource,
        json:    json,
        debug:   debug,
        quiet:   quiet,
        verbose: verbose,
      )
      if res_version_info.empty?
        status_hash(resource, "error", ["Unable to get versions"], verbose: verbose)
      else
        res_version_info
      end
    end
  end.compact_blank

  Homebrew.failed = true if resource_version_info.any? { |info| info[:status] == "error" }
end
```

mohammadzainabbas commented on Library/Homebrew/livecheck/livecheck.rb:
```diff
@@ -443,47 +492,68 @@ def print_latest_version(info, verbose:, ambiguous_cask: false)
 
-      case formula_or_cask
+      case package_or_resource
+      when Resource
+        urls << package_or_resource.url
```
Done ✅

mohammadzainabbas commented on Library/Homebrew/livecheck/livecheck.rb:
```diff
@@ -561,7 +631,200 @@ def use_homebrew_curl?(formula_or_cask, url)
+
+      livecheck = resource.livecheck
+      livecheck_url = has_livecheckable ? livecheck.url : resource.url
+      livecheck_regex = livecheck&.regex
+      livecheck_strategy = livecheck&.strategy
+      livecheck_strategy_block = livecheck&.strategy_block
```
Initially when I implemented the `resource_version` method, I was getting an error here for resources which do not have any `livecheck` blocks. That's why I used safe navigation operator here to avoid getting error in those cases. 

But now, I have checked it again after removing the guards and it works. Not sure why it's working now without the safe navigation operator.

samford replied:
The `@livecheck` instance variable is set to a new `Livecheck` object when a `Resource` is initialized (regardless of whether a `livecheck` block is present in the resource) and `Livecheck` objects also have default instance variable values, so the only way I can imagine these calls failing is if `resource` isn't a `Resource` object. If I had to guess, maybe there weren't sufficient safeguards to protect against that situation at some point (or maybe something was broken while making tweaks to debug/test). Either way, it seems fine now.

mohammadzainabbas replied:
> If I had to guess, maybe there weren't sufficient safeguards to protect against that situation at some point (or maybe something was broken while making tweaks to debug/test)

Yes, I noticed that `@livecheck = nil` was changed to `@livecheck = Livecheck.new(self)` in https://github.com/Homebrew/brew/pull/13753, which fixed the issue I was having initially.

mohammadzainabbas commented on Library/Homebrew/livecheck/livecheck.rb:
```diff
@@ -561,7 +631,200 @@ def use_homebrew_curl?(formula_or_cask, url)
+        res_current = resource.version
+        res_latest = Version.new(match_version_map.values.max_by { |v| LivecheckVersion.create(resource, v) })
+
+        return status_hash(resource, "error", ["Unable to get versions"], verbose: verbose) if res_latest.to_s.blank?
```
Oh yeah, you are right. Fixed ✅

mohammadzainabbas commented on Library/Homebrew/livecheck/livecheck.rb:
```diff
@@ -561,7 +631,200 @@ def use_homebrew_curl?(formula_or_cask, url)
+
+        return status_hash(resource, "error", ["Unable to get versions"], verbose: verbose) if res_latest.to_s.blank?
+
+        is_newer_than_upstream = res_current > res_latest
+        is_outdated = (res_current != res_latest) && !is_newer_than_upstream
```
Okay 👍

mohammadzainabbas commented on Library/Homebrew/livecheck/livecheck.rb:
```diff
@@ -561,7 +631,200 @@ def use_homebrew_curl?(formula_or_cask, url)
+        end
+        res_livecheck[:regex] = regex.inspect if regex.present?
+        res_livecheck[:cached] = true if strategy_data[:cached] == true
+        resource_version_info[:meta][:livecheck] = res_livecheck
```
> ```json
> {
>   "resource": "pkg-config-wrapper",
>   "version": {
>     "current": "0.2.11",
>     "latest": "0.2.12",
>     "outdated": true,
>     "newer_than_upstream": false
>   },
>   "meta": {
>     "url": {
>       "original": "https://github.com/influxdata/pkg-config/archive/refs/tags/v0.2.11.tar.gz",
>       "processed": "https://github.com/influxdata/pkg-config.git"
>     },
>     "livecheckable": true,
>     "livecheck": {
>       "url": {
>         "original": "url",
>         "symbol": "url"
>       },
>       "strategy": "Git",
>       "strategies": [
>         "Git",
>         "PageMatch"
>       ],
>       "regex": "/^v?(\\d+(?:\\.\\d+)+)$/i"
>     }
>   }
> }
> ```

I failed to reproduce this issue. It was working fine locally. i.e: I wasn't getting `resource_version_info[:meta][:livecheck][:url][:original] or [symbol] = 'url'`

```json
{
  "resource": "pkg-config-wrapper",
  ...
  "meta": {
    ...
    "livecheckable": true,
    "livecheck": {
      "url": {
        "original": "url",
        "symbol": "url"
      },
      "strategy": "Git",
      ...
    }
  }
}

``` 

However, I have fixed the issues with json output as suggested. Now, it's almost identical to `latest_version`'s logic.

mohammadzainabbas commented on Library/Homebrew/livecheck/livecheck.rb:
```diff
@@ -561,7 +631,200 @@ def use_homebrew_curl?(formula_or_cask, url)
+        Homebrew.failed = true
+        if json
+          status_hash(resource, "error", [e.to_s], verbose: verbose)
+        elsif onoe "#{Tty.blue}#{resource.name}#{Tty.reset}: #{e}"
```
Updated `#resource_version` method and it's call ✅

mohammadzainabbas commented on Library/Homebrew/livecheck/livecheck.rb:
```diff
@@ -730,6 +994,7 @@ def latest_version(
         end
 
         version_info = {
+          # latest: {},
```
Oops! my bad. Turns out I had it removed locally but didn't commit the change.

mohammadzainabbas commented on Library/Homebrew/livecheck/livecheck.rb:
```diff
@@ -561,7 +631,200 @@ def use_homebrew_curl?(formula_or_cask, url)
+        debug:    T::Boolean,
+      ).returns(Hash)
+    }
+    def resource_version(
```
Oh okay, done ✅

mohammadzainabbas commented on Library/Homebrew/livecheck/skip_conditions.rb:
```diff
@@ -33,8 +33,8 @@ def formula_or_cask_skip(formula_or_cask, livecheckable, full_name: false, verbo
 
-        skip_message = if formula_or_cask.livecheck.skip_msg.present?
-          formula_or_cask.livecheck.skip_msg
+        skip_message = if package_or_resource.livecheck&.skip_msg.present?
```
Okay!

mohammadzainabbas commented on Library/Homebrew/livecheck/skip_conditions.rb:
```diff
@@ -45,10 +45,10 @@ def formula_or_cask_skip(formula_or_cask, livecheckable, full_name: false, verbo
         end
 
-        return {} if !formula_or_cask.livecheck.skip? && skip_message.blank?
+        return {} if !package_or_resource.livecheck&.skip? && skip_message.blank?
```
✅

mohammadzainabbas commented on Library/Homebrew/test/livecheck/livecheck_spec.rb:
```diff
@@ -123,30 +150,56 @@
+      expect(livecheck.livecheck_url_to_string(head_url, r_livecheck_url)).to eq(head_url)
+
+      r_livecheck_url.livecheck.url(:stable)
+      expect(livecheck.livecheck_url_to_string(stable_url, r_livecheck_url)).to eq(stable_url)
```
Yeah, I was getting errors for the `:head` / `:stable` after modification. 

I asked @nandahkrishna once that `:head` and `:stable` are implemented in `Formula` as a `SoftwareSpec`. And (to my knowledge) if we want to have the same for resources, we have to update the `Resource` DSL. And he told me:

> `Resource`'s `:url` is considered `:stable` by default.

And I took it in a sense that if I got `:stable` symbol, I can return `:url` for resource in `#livecheck_url_to_string` method. (kindly do correct me if I am mistaken --- which I think I most probably am)

Anyways, I have removed tests for `:head` and `:stable` for the time being.

samford commented on Library/Homebrew/livecheck/livecheck.rb:
```diff
@@ -443,47 +492,68 @@ def print_latest_version(info, verbose:, ambiguous_cask: false)
+        package_or_resource.url&.to_s if package_or_resource.is_a?(Cask::Cask) || package_or_resource.is_a?(Resource)
       when :head, :stable
-        formula_or_cask.send(livecheck_url)&.url if formula_or_cask.is_a?(Formula)
+        # Since resource's "url" is considered "stable" by default.
+        # And some resources may contain in "head" block as well
+        package_or_resource.send(:url)&.to_s if package_or_resource.is_a?(Resource)
```
> I asked @nandahkrishna once that :head and :stable are implemented in Formula as a SoftwareSpec. And (to my knowledge) if we want to have the same for resources, we have to update the Resource DSL. And he told me:
> 
> > Resource's :url is considered :stable by default.
>
> And I took it in a sense that if I got :stable symbol, I can return :url for resource in #livecheck_url_to_string method. (kindly do correct me if I am mistaken --- which I think I most probably am)
>
> Anyways, I have removed tests for :head and :stable for the time being.

I'm of the opinion that `url :head` and `url :stable` in resource `livecheck` blocks should work the same as casks, where `#livecheck_url_to_string` doesn't handle these symbols (i.e., implicitly returning `nil`). At least at present, neither casks nor resources respond to `#head` or `#stable` methods and they're not part of their DSLs, so using those values in the `livecheck` block shouldn't work.

At least to me, we shouldn't be trying to quietly correct this issue for users by substituting the resource `url`, as this will lead people to think that `url :head` and `url :stable` are valid and we don't want this mistake to propagate through homebrew/core. Ideally, we should enforce valid symbols with a Rubocop but that's something for a later PR because currently the related Rubocops only work for formulae (not casks or resources) and fixing that is a significant undertaking. [I have some in-progress work but it's currently blocked while I wait for #13233 to be finished/merged.]

That said, at least from my perspective, this comment and statement should be removed from `#livecheck_url_to_string` (so the method will implicitly return `nil` when these symbols are used in a resource `livecheck` block, same as casks).

@nandahkrishna What do you think about this?

nandahkrishna replied:
After some thought, I agree with Sam on this – let's remove this comment and statement for consistency with the Cask behaviour. So in short, we will get the URL only if we use `:url`, and not `:stable` or `:head`.

I was earlier concerned about `:stable` being valid in some sense (resources included in a formula are most likely "stable" by default), but the explanation that resources do not respond to the `head` and `stable` methods convinces me that we shouldn't quietly succeed by returning the `url`.

mohammadzainabbas replied:
Alright! I have removed this. Thank you @samford and @nandahkrishna for the explanation.

samford commented on Library/Homebrew/livecheck/livecheck.rb:
```diff
@@ -243,11 +247,37 @@ def run_checks(
           )
         end
 
+        # Check current and latest resources (if "--resources" flag is given)
+        # Only check current and latest versions if we have resources to check against
+        check_for_resources = check_resources && formula_or_cask.is_a?(Formula) && formula_or_cask.resources.present?
+        if check_for_resources
+          resource_version_info = formula_or_cask.resources.map do |resource|
+            res_skip_info ||= SkipConditions.skip_information(resource, verbose: verbose)
+            if res_skip_info.present?
+              res_skip_info
+            else
+              res_version_info = resource_version(
+                resource,
+                json:    json,
+                debug:   debug,
+                quiet:   quiet,
+                verbose: verbose,
+              )
+              if res_version_info.empty?
+                status_hash(resource, "error", ["Unable to get versions"], verbose: verbose)
+              else
+                res_version_info
+              end
+            end
+          end.compact_blank
+        end
+
```
This code should come after the main formula checks, as resource checks are secondary to the main formula check. Currently, because this code comes before the main check, the debug output shows the resource information before the formula information, which isn't correct. For example, with my modified version of `influxdb` I'm seeing:

```
Resource:         pkg-config-wrapper
Livecheckable?:   Yes

URL (url):        https://github.com/influxdata/pkg-config/archive/refs/tags/v0.2.11.tar.gz
URL (processed):  https://github.com/influxdata/pkg-config.git
Strategy:         Git
Regex:            /^v?(\d+(?:\.\d+)+)$/i

Matched Versions:
0.2.1, 0.1.0, 0.1.1, 0.2.0, 0.2.10, 0.2.11, 0.2.12, 0.2.2, 0.2.3, 0.2.4, 0.2.5, 0.2.6, 0.2.7, 0.2.8, 0.2.9
Formula:          influxdb
Livecheckable?:   Yes

URL (stable):     https://github.com/influxdata/influxdb.git
Strategy:         Git
Regex:            /^v?((?!9\.9\.9)\d+(?:\.\d+)+)$/i

Matched Versions:
1.8.0, 0.0.1, 0.0.2, 0.0.3, 0.0.4, 0.0.5, 0.0.6, 0.0.7, 0.0.8, 0.0.9, 0.1.0, 0.10.0, 0.10.1, 0.10.2, 0.10.3, 0.11.0, 0.11.1, 0.12.0, 0.12.1, 0.12.2, 0.13.0, 0.2.0, 0.3.0, 0.3.1, 0.3.2, 0.4.0, 0.4.1, 0.4.2, 0.4.3, 0.4.4, 0.5.0, 0.5.1, 0.5.10, 0.5.11, 0.5.12, 0.5.2, 0.5.3, 0.5.4, 0.5.5, 0.5.6, 0.5.7, 0.5.8, 0.5.9, 0.6.0, 0.6.1, 0.6.2, 0.6.3, 0.6.4, 0.6.5, 0.7.0, 0.7.1, 0.7.2, 0.7.3, 0.8.0, 0.8.1, 0.8.2, 0.8.3, 0.8.4, 0.8.5, 0.8.6, 0.8.7, 0.8.8, 0.9.0, 0.9.1, 0.9.2, 0.9.2.1, 0.9.3, 0.9.4, 0.9.4.1, 0.9.4.2, 0.9.5, 0.9.5.1, 0.9.6, 0.9.6.1, 1.0.0, 1.0.1, 1.0.2, 1.1.0, 1.1.1, 1.1.2, 1.1.3, 1.1.4, 1.1.5, 1.10.0, 1.2.0, 1.2.1, 1.2.2, 1.2.3, 1.2.4, 1.3.0, 1.3.1, 1.3.2, 1.3.3, 1.3.4, 1.3.5, 1.3.6, 1.3.7, 1.3.8, 1.3.9, 1.4.0, 1.4.1, 1.4.2, 1.4.3, 1.5.0, 1.5.1, 1.5.2, 1.5.3, 1.5.4, 1.5.5, 1.6.0, 1.6.1, 1.6.2, 1.6.3, 1.6.4, 1.6.5, 1.6.6, 1.7.0, 1.7.1, 1.7.10, 1.7.11, 1.7.2, 1.7.3, 1.7.4, 1.7.5, 1.7.6, 1.7.7, 1.7.8, 1.7.9, 1.8.1, 1.8.10, 1.8.2, 1.8.3, 1.8.4, 1.8.5, 1.8.6, 1.8.7, 1.8.8, 1.8.9, 1.9.0, 1.9.1, 1.9.2, 1.9.3, 1.9.4, 1.9.5, 1.9.6, 1.9.7, 1.9.8, 2.0.0, 2.0.1, 2.0.2, 2.0.3, 2.0.4, 2.0.5, 2.0.6, 2.0.7, 2.0.8, 2.0.9, 2.1.0, 2.1.1, 2.2.0, 2.3.0, 2.4.0

influxdb: 2.4.0 ==> 2.4.0
  pkg-config-wrapper: 0.2.11 ==> 0.2.12
  ui-assets: skipped - Update manually
```

mohammadzainabbas replied:
I moved this before the `SkipConditions` logic for `Formula` as per @nandahkrishna's suggestion. i.e: we should still get the resources' info in case `Formula` was skipped.

nandahkrishna replied:
@mohammadzainabbas I believe there's a misunderstanding here. I think:
* If a Formula is skipped, its resources don't matter (we can ignore checking them).
* If a Formula's `livecheck` errors out, we should still attempt to show resource versions.

The resource outputs should always be printed after the formula output.

mohammadzainabbas replied:
Oh okay, I have updated the code accordingly.

samford commented on Library/Homebrew/livecheck/livecheck.rb:
```diff
@@ -770,6 +842,202 @@ def latest_version(
+      resource_version_info = {}
+
+      livecheck = resource.livecheck
+      livecheck_url = has_livecheckable ? livecheck.url : resource.url
```
```suggestion
      livecheck_url = livecheck.url
```

As seen in `#latest_version`, the `livecheck_url` value should simply correspond to `livecheck.url` (i.e., we store the value in a variable to avoid repeatedly making a function call wherever `livecheck.url` is used). When it's set, it's the symbol/string from the `livecheck` block, otherwise it's the default `nil` value. That is to say, this shouldn't be falling back to the resource `url` because we handle that in subsequent code (i.e., when defining the `urls` variable).

mohammadzainabbas replied:
Yes, you are right. I had to add this check initially when `@livecheck` was `nil` in `Resource` DSL (before https://github.com/Homebrew/brew/pull/13753) to avoid getting error on resources which had no `livecheck` blocks.

samford commented on Library/Homebrew/livecheck/livecheck.rb:
```diff
@@ -443,47 +493,69 @@ def print_latest_version(info, verbose:, ambiguous_cask: false)
+      info.each do |r_info|
+        if r_info[:status] && r_info[:messages]
+          SkipConditions.print_skip_information(r_info)
+          Homebrew.failed = true
```
The presence of `:status`/`:messages` values doesn't always indicate a failure condition. For example, these values would be set for a skipped resource but that's not a failure state. The previous use of `Homebrew.failed = true` was fine from my perspective (i.e., it mimicked the behavior in `#run_checks`) but this is setting the failure state for more than it should.

@nandahkrishna Can you elaborate on which scenarios you were thinking should be treated as a failure state with regard to resources? This may help to clarify where `Homebrew.failed = true` should be used outside of the `rescue` blocks.

nandahkrishna replied:
I wanted to make sure that an `Unable to get versions` for resources would be treated as a failure state. Prior to this change, we did not have an error status code if a formula's `livecheck` succeeded but a resource's failed. This was the only thing I had in mind so far about failure states for resources.

mohammadzainabbas replied:
Just a naive idea, we can use `:status` to identify if the resource was skipped or was it due to some failed state, and mark `Homebrew.failed = true` in that case. This way we won't get false status code in case of a skipped resource. @samford and @nandahkrishna: what are your thoughts on this ?

samford replied:
It looks like GitHub lost both of our comments (my response and your reply to it), so I'll quote it from the email/DM:

> > However, this should arguably be handled somewhere after #resource_version is called in #run_checks
>
> Is this what you had in mind ?
>
> ```ruby
> check_for_resources = check_resources && formula_or_cask.is_a?(Formula) && formula_or_cask.resources.present?
> if check_for_resources
>   resource_version_info = formula_or_cask.resources.map do |resource|
>     res_skip_info ||= SkipConditions.skip_information(resource, verbose: verbose)
>     if res_skip_info.present?
>       res_skip_info
>     else
>       res_version_info = resource_version(
>         resource,
>         json:    json,
>         debug:   debug,
>         quiet:   quiet,
>         verbose: verbose,
>       )
>       if res_version_info.empty?
>         status_hash(resource, "error", ["Unable to get versions"], verbose: verbose)
>       else
>         res_version_info
>       end
>     end
>   end.compact_blank
> Homebrew.failed = true if resource_version_info.any? { |info| info[:status] == "error" }
> end
> ```

Yes, that's what I was thinking. From my limited testing, it appears to work as expected and would handle Nanda's use case. Feel free to push a commit and I'll take a final pass through this PR.

mohammadzainabbas replied:
I have pushed the change ✅

samford reviewed (commented):
I'll take a more thorough pass through this in the evening (EDT) but here are a few remaining items to address/discuss. It feels like we're very close to finishing this off, though.

samford reviewed (approved):
I took a final pass through this and pushed some commits to mostly tweak comments/text and some code style in areas. The only functional change I made was to remove the added `resource` parameter from `#print_latest_version`, as I found that its purpose could be met by using `info[:resource].present?` instead.

I did full runs across homebrew/core and homebrew/cask, comparing the JSON output on the `master` branch and this PR branch and didn't observe any issues. I also did some manual testing with the `--resource` option set in normal, debug, JSON, and verbose JSON scenarios and everything I tested looked fine. With that in mind, I'm good with this being merged after @nandahkrishna has had a chance to review and approve it.

-----

As mentioned before, there's still room for improvement here but our time is very limited, so I plan to tackle some of it as part of planned refactoring. There's also work that will need to be done to make use of this feature (e.g., creating a comprehensive way of bumping resource versions, integrating it with `brew bump-formula-pr`, etc.). Properly handling resource versions was arguably going to be the challenging part of this project (i.e., the newest available upstream version isn't always appropriate for a given package). However, we're almost at the end of our allotted time, so this is something for the future.

nandahkrishna reviewed (approved):
Thank you very much Sam for the comprehensive review. As discussed, this feature is still far from complete and the code definitely needs to be refactored, but we can take care of this after GSoC ends (I'm very eager to contribute to this!).

I went through the code and pushed a minor change to modify a couple of comments. Some brief testing seems to indicate to me that everything works as intended – so I'm happy to merge this PR. Once this is merged, we can start adding `livecheck` blocks to resources in a few Formulae in homebrew-core (just like Mohammad's open PR there).

Thanks for your work on this, @mohammadzainabbas.

nandahkrishna commented:
The failing tests here are a known issue (discussed on Slack) and unrelated to the changes made in this PR. We can merge once the issue is fixed.

mohammadzainabbas commented:
> As discussed, this feature is still far from complete and the code definitely needs to be refactored, but we can take care of this after GSoC ends (I'm very eager to contribute to this!).

Thank you @samford and @nandahkrishna for your valuable feedbacks. I would very much like to take part in code refactoring as well (after GSoC) and keep contributing to __Homebrew__ code base (esp. `livecheck` module).

nandahkrishna commented:
The issue seems fixed, so I'll be merging this.
